contracts/exchange/test/match_orders.ts

```diff
@@ -1270,37 +1228,261 @@ describe('matchOrders', () => {
+                    rightMakerFee: TokenType.ERC20B,
+                    leftTakerFee: TokenType.ERC20A,
+                    rightTakerFee: TokenType.ERC20B,
+                },
+                {
+                    description: 'Paying maker and taker fees with same tokens being bought.',
+                    leftMaker: TokenType.ERC20A,
+                    rightMaker: TokenType.ERC20B,
+                    leftMakerFee: TokenType.ERC20B,
+                    rightMakerFee: TokenType.ERC20A,
+                    leftTakerFee: TokenType.ERC20B,
+                    rightTakerFee: TokenType.ERC20A,
+                },
+                {
+                    description: 'Buy an ERC721 then use it to pay maker fees.',
+                    leftMaker: TokenType.ERC20A,
+                    rightMaker: TokenType.ERC721A,
+                    leftMakerFee: TokenType.ERC721A,
+                    rightMakerFee: TokenType.ERC20B,
+                    leftTakerFee: TokenType.ERC20C,
+                    rightTakerFee: TokenType.ERC20C,
+                },
+                {
+                    description: 'Buy an ERC721 then use it to pay maker fee (the other way).',
+                    leftMaker: TokenType.ERC721A,
+                    rightMaker: TokenType.ERC20A,
+                    leftMakerFee: TokenType.ERC20B,
+                    rightMakerFee: TokenType.ERC721A,
+                    leftTakerFee: TokenType.ERC20C,
+                    rightTakerFee: TokenType.ERC20C,
+                },
+                {
+                    description: 'Swapping ERC721s then using them to pay maker fees.',
+                    leftMaker: TokenType.ERC721A,
+                    rightMaker: TokenType.ERC721B,
+                    leftMakerFee: TokenType.ERC721B,
+                    rightMakerFee: TokenType.ERC721A,
+                    leftTakerFee: TokenType.ERC20A,
+                    rightTakerFee: TokenType.ERC20A,
+                },
+                {
+                    description: 'Double-spend by trying to pay maker fees with sold ERC721 token (fail).',
+                    leftMaker: TokenType.ERC721A,
+                    rightMaker: TokenType.ERC721B,
+                    leftMakerFee: TokenType.ERC721A,
+                    rightMakerFee: TokenType.ERC721A,
+                    leftTakerFee: TokenType.ERC20A,
+                    rightTakerFee: TokenType.ERC20A,
+                    shouldFail: true,
+                },
+            ];
+
+            let erc721TokenIdsByOwner: ERC721TokenIdsByOwner;
+            let nameToERC20Tokens: { [name: string]: string };
+            let nameToERC721Tokens: { [name: string]: string };
+
+            function getAssetData(tokenType: TokenType, ownerAddress: string): string {
+                const encodeERC20AssetData = assetDataUtils.encodeERC20AssetData;
+                const encodeERC721AssetData = assetDataUtils.encodeERC721AssetData;
+                if (tokenType in nameToERC20Tokens) {
```

Comment:
Nit: the convention we use throughout the rest of the codebase is `if (nameToERC20Tokens[tokenType] !== undefined)`